kong/router.lua

```diff
@@ -768,49 +750,35 @@ function _M.new(routes)
 
           if matched_route then
             local upstream_host
-            local upstream_uri   = req_uri
+            local upstream_uri
             local upstream_url_t = matched_route.upstream_url_t
             local matches        = ctx.matches
 
-            -- URI stripping logic
 
-            local uri_root = req_uri == "/"
-
-            if not uri_root and matched_route.strip_uri
-               and matches.stripped_uri
-            then
-              upstream_uri = matches.stripped_uri
-            end
+            -- Path construction
 
-            -- uri trailing slash logic
+            -- if we do not have a path-match, then the postfix is simply the
+            -- incoming path, without the initial slash
+            local request_postfix = matches.stripped_uri or sub(req_uri, 2, -1)
+            local upstream_base = upstream_url_t.path
 
-            local upstream_url_path = upstream_url_t.path
-            local upstream_url_file = upstream_url_t.file
-
-            if upstream_url_path and upstream_url_path ~= "/" then
-              if upstream_uri ~= "/" then
-                upstream_uri = upstream_url_file .. upstream_uri
+            if matched_route.strip_uri then
+              -- we drop the matched part, replacing it with the upstream path
+              if sub(upstream_base, -1, -1) == "/" and sub(request_postfix, 1, 1) == "/" then
```

Comment:
The first statement of this condition (`if sub(upstream_base, -1, -1) == "/"`) could be avoided if we relied on `upstream_url_t.file`, which we sanitize only once, [here](https://github.com/Kong/kong/blob/fix/path-handling/kong/router.lua#L265-L266), to ensure that it does not contain a trailing slash.

And when `upstream_base == "/"` (initialized to `file` instead of `path`), we should then do: `upstream_uri = request_postfix`. That would be cheaper. 

Besides, it seems like the third argument in the first `sub()` call is unnecessary, but that's irrelevant if superseded by the above.

Reply:
style: mind keeping the line length below 80 characters please?

Reply:
Considering this code will be moved out of this pre-access code, into post-access, I'd rather not as it seems a premature optimization doing that now. Once a follow up PR puts this in the proper place, we can optimize it there imo.

That said, afaik, the `file` thing is only used here in the router, should it be removed? Once the path construction moves to after-access, we cannot rely on it any more.

Reply:
It wasn't clear that there would be a subsequent PR. In this case, I'd rather see the final result of these changes at once, in the same PR (separate commits is fine). I'd argue that aiming towards commit atomicity does mean that each commit should be written optimally: say we end up reverting commit number 2, now commit number 1 does not include the optimizations that were only added in 2.

> That said, afaik, the file thing is only used here in the router, should it be removed? 

When a variable becomes stale (and only then), please, by all means, remove it!